infra/src/main/kotlin/fi/fta/geoviite/infra/geocoding/GeocodingDao.kt

```diff
@@ -57,24 +54,32 @@ class GeocodingDao(
         val sql =
             """
             select
-              tn.official_id as tn_official_id,
-              tn.row_id as tn_row_id,
-              tn.row_version as tn_row_version,
-              rl.row_id as rl_row_id,
-              rl.row_version as rl_row_version,
-              array_agg(kmp.row_id order by kmp.row_id, kmp.row_version) 
-                filter (where kmp.row_id is not null) as kmp_row_ids,
-              array_agg(kmp.row_version order by kmp.row_id, kmp.row_version) 
-                filter (where kmp.row_id is not null) as kmp_row_versions
+              tn.id as tn_id,
+              tn.design_id as tn_design_id,
+              tn.draft as tn_draft,
+              tn.version as tn_version,
+              rl.id as rl_id,
+              rl.design_id as rl_design_id,
+              rl.draft as rl_draft,
+              rl.version as rl_version,
+              kmp_ids,
+              kmp_design_ids,
+              kmp_drafts,
+              kmp_versions
             from layout.track_number_in_layout_context(:publication_state::layout.publication_state, :design_id) tn
               left join
                 layout.reference_line_in_layout_context(:publication_state::layout.publication_state, :design_id)
-                  rl on rl.track_number_id = tn.official_id
-              left join layout.km_post_in_layout_context(:publication_state::layout.publication_state, :design_id)
-                kmp on kmp.track_number_id = tn.official_id
-                and kmp.state = 'IN_USE'
-            where ((:tn_id::int is null and tn.state != 'DELETED') or :tn_id = tn.official_id)
-            group by tn.official_id, tn.row_id, tn.row_version, rl.row_id, rl.row_version
+                  rl on rl.track_number_id = tn.id
+              left join lateral (
+                select
+                  coalesce(array_agg(kmp.id order by id), '{}') as kmp_ids,
+                  coalesce(array_agg(kmp.design_id order by id), '{}') as kmp_design_ids,
+                  coalesce(array_agg(kmp.draft order by id), '{}') as kmp_drafts,
+                  coalesce(array_agg(kmp.version order by id), '{}') as kmp_versions
+                  from layout.km_post_in_layout_context(:publication_state::layout.publication_state, :design_id) kmp
+                  where kmp.track_number_id = tn.id and kmp.state = 'IN_USE'
+              ) kmp on (true)
+            where ((:tn_id::int is null and tn.state != 'DELETED') or :tn_id = tn.id)
```

Comment:
Näitä `coalesce(array_agg(..), '{}')`-juttuja tulee aika paljon, ei sinällään varsinaisesta välttämättömästä tarpeesta tehdä niitä kaikkia tässä pullarissa, mutta siksi, koska postgres ei osaa päätellä kovin paljoa siitä, mitä sarakkeita voi jättää pois tuosta `group by`-osiosta. Siksi päädyin siihen, että heitetään grouppaukset kokonaan pois ja muutetaan näiden hakujen rakenne vaan sellaiseksi, että 1:n-suhteiset haut tehdään eksplisiittisesti omina erillisinä asioinaan.

Periaatteessa siis postgressissä `coalesce(array_agg(...), '{}')` tarkoittaa samaa kuin Kotlinissa ` ?: listOf()`, ja tuo merkkijonomuoto sattuu olemaan oikea tapa sanoa asia lyhyesti niin, että tyyppijärjestelmäkin sen hyväksyy.